File: src/main/java/leetcode/ProductExceptSelf.java
package leetcode;

public class ProductExceptSelf {
  public int[] productExceptSelf(int[] nums) {
    int[] products = new int[nums.length];
    int prefix = 1;
    for (int i = 0; i < nums.length; i++) {
      products[i] = prefix;
      prefix *= nums[i];
    }
    int suffix = 1;
    for (int i = nums.length - 1; i >= 0; i--) {
      products[i] *= suffix;
      suffix *= nums[i];
    }
    return products;
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/ProductExceptSelf.kt
package com.giwankim.leetcode

class ProductExceptSelf {
    fun productExceptSelf(nums: IntArray): IntArray {
        val products = IntArray(nums.size)
        var prefix = 1
        for (i in nums.indices) {
            products[i] = prefix
            prefix *= nums[i]
        }
        var suffix = 1
        for (i in nums.indices.reversed()) {
            products[i] *= suffix
            suffix *= nums[i]
        }
        return products
    }
}
